merge.py

```diff
@@ -1,98 +1,303 @@
-  mvn_success = call(['mvn', 'clean', 'verify'], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  if not mvn_success == 0:
-    post_error('mvn clean verify against HEAD + PR#{} failed. Not merging.'.format(pr), pr)
-    return False
-
-  # git push (COMMENTED UNTIL MERGEBOT HAS PERMISSIONS)
-  #push_success = call(['git', 'push', 'apache', 'HEAD:master'], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  #if not push_success == 0:
-  #  post_error('Git push failed. Please try again.', pr)
-  #  return False
-  return True
-
-def set_up():
-  if not call(['mkdir', '/usr/local/google/home/jasonkuster/tmp']) == 0:
-    return False
-  return True
-
-
-def clean_up():
-  if not call(['rm', '-rf', '/usr/local/google/home/jasonkuster/tmp']) == 0:
-    return False
-  return True
 
+
+import abc
+import logging
+import os
+import shlex
+from subprocess import check_call, check_output, STDOUT
+from threading import Thread
+
+
+APACHE_GIT = 'https://git-wip-us.apache.org/repos/asf/{repo}.git'
+GITHUB_REPO_URL = 'https://github.com/{org}/{repo}.git'
+TMP_DIR_FMT = '/tmp/{dir}'
+
+
+def create_merger(config, work_queue):
+    """create_merger creates a merger of the type specified in the config.
+
+    Args:
+        config: A dictionary containing repository configuration.
+        work_queue: A Queue.Queue where new work items will be queued.
+    Returns:
+        Merger of type specified in configuration
+    Raises:
+        AttributeError: if passed an unsupported SCM type.
+    """
+    if config['scm_type'] == 'github':
+        return GitMerger(config, work_queue)
+    raise AttributeError(
+        'Unsupported SCM type: {}.'.format(config['scm_type']))
+
+
+class Merger(Thread):
+    """Merger is the base class for all mergers.
+    """
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, config, work_queue):
+        self.config = config
+        l = logging.getLogger('{name}_merge_logger'.format(
+            name=self.config['name']))
+        log_fmt = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+        f = logging.Formatter(log_fmt)
+        filename = '{name}_merger_log.txt'.format(name=config['name'])
+        h = logging.FileHandler(os.path.join('log', filename), mode='w')
+        h.setFormatter(f)
+        l.addHandler(h)
+        l.setLevel(logging.INFO)
+        self.main_logger = l
+        self.work_queue = work_queue
+        Thread.__init__(self)
+
+    @abc.abstractmethod
+    def run(self):
+        """run spins forever, merging things off of the work queue.
+        """
+        return
+
+
+class GitMerger(Merger):
+    """GitMerger merges Git pull requests.
+    """
+
+    def run(self):
+        """run spins forever, merging things off of the work queue.
+
+        Raises:
+            AttributeError: if work items are not of type github_helper.GithubPR
+        """
+        while True:
+            pr = self.work_queue.get()
+            pr_num = pr.get_num()
+            self.main_logger.info(
```

Comment:
The idea is for the main logger to log to the *_merge.txt file as an overview, and then for each PR merge to have its own log file. This could be accomplished any of a number of different ways, but this is one that works. Passing a mock logger for testing can be accomplished by mocking the logger itself.